### ISSUE TITLE
"async & performance" event loop & job queue

### ISSUE THREAD

**Issue by riskers** [getify/You-Dont-Know-JS#882]
when I read "async & performance" section , I'm so puzzled about `event loop` and `job queue`:

* What is the order of execution `event loop` and `job queue` ? What is their relationship?
* `job queue` is same as `task queue`?

Do you have any articles to discuss such issues ? thanks..

**Comment by deltaidea** [comment]
This is what you're looking for: [Philip Roberts: What the heck is the event loop anyway?](https://www.youtube.com/watch?v=8aGhZQkoFbQ)

**Comment by riskers** [comment]
@deltaidea 

for example:

```js
var start = new Date().getTime()

setTimeout(function(){
    console.log(1)
},3000);

while(new Date().getTime() - start < 3000){
}
console.log(2)
```

why `1` and `2` log at the same time?

Isn't put `setTimeout` in event loop then hold while loop 3000ms, then log 2. 
After, enter event loop find setTimeout then delay 3000ms , log 1?

**Comment by deltaidea** [comment]
Seriously, just watch the talk in my first reply. :)

```JS
// Remember initial time.
var start = new Date().getTime()

// This code runs inside JS engine which can use other parts of the browser.
// `window.setTimeout` is a function that invokes a browser API,
// kinda like executing an external program.
// That browser API is not a part of the engine, it doesn't obey the limit
// to only run one thing at a time.
// `setTimeout` asks the browser to independently wait for 3 seconds and then
// push this function into the job queue.
setTimeout(function(){
    console.log(1)
},3000);

// This `while` loop will run inside the JS engine for 3 seconds,
// doing nothing in each iteration and constantly checking current time.
while(new Date().getTime() - start < 3000){
}

// This line will execute immediately after the `while` loop has finished.
console.log(2)

// `while` and other loops have nothing to do with the JS event loop.

// The event loop is the thing in the browser that checks if the JS engine
// is idle and when it is - the loop gives it the first item in the queue to run.
// The engine will not be idle until the last line here is executed.
// Meanwhile, the browser will eventually put `console.log(1)` into
// the job queue, maybe even before the `while` loop is finished.
// But `console.log(1)` will only run once the engine has nothing to do.
// Once the engine finishes running the last line - `console.log(2)`,
// the event loop will get `console.log(1)` from the task queue and give it
// to the engine to run. That's why you'll always see `2` before `1`.
```

**Comment by riskers** [comment]
@deltaidea thx 👍 

In addition, I also confused with some terminology such as `Promise`, `macrotasks`, `microtasks` from some articles. How to understand them? Thank you very much.

**Comment by deltaidea** [comment]
Personally I have no idea what you're talking about. It's probably related to a particular library used in those articles.

I think this issue may be closed now.

**Comment by derek-knox** [comment]
@deltaidea @riskers

There can be  more than one Job Queue (https://www.ecma-international.org/ecma-262/6.0/#table-26)

tl;dr precedence
1. Microtask Queue (promises)
2. Event/Callback Queue (setTimeout, setInterval, etc.)

**Comment by alexharty** [comment]
How do I unsub from this thread? I already logged into github and hit
unsubscribe but I still get these.

**Comment by jinbeomhong** [comment]
For the record. You can think of job queue as microtask queue. They are different thing. For more information, it is helpful to see https://youtu.be/cCOL7MC4Pl0
